doc/design/queue_api.md

```diff
@@ -0,0 +1,178 @@
+|   cpu: 6       |    |   cpu: 3       | 
+|   memory: 15Gi |    |   memory: 12Gi | 
+------------------    ------------------ 
+---------------    ---------------
+| Queue-1     |    | Queue-2     |
+|   Weight: 2 |    |   Weight: 4 |
+---------------    ---------------
+
+After policy scheduling:
+--------------------    ---------------------
+| Queue-1          |    | Queue-2           |
+|    Weight: 2     |    |    Weight: 4      |
+|                  |    |                   |
+|    Allocated:    |    |    Allocated:     |
+|      cpu: 3      |    |      cpu: 6       |
+|      memory: 9Gi |    |      memory: 18Gi | 
+--------------------    ---------------------
+```
+
+Policy format scheduler results as `QueueInfo` and transfers to Preemption for next step.
+
+```go
+type QueueInfo struct {
+   // The name of the queue
+	name  string
+	// Queue information contains Deserved/Allocated/Used/Preempting
+	queue *Queue
+	// Running pods under this queue
+	pods  map[string]*Pod
+}
+```
+
+### Preemption
+
+Preemption is used to reclaim resource for overused(`Deserved` < `Allocated`) queue. The following status will be guaranteed after preemption done.
+
+* The cluster won't be overused. 
+* Each queue must meet `Deserved = Allocated`, it means the queue will get all resources allocated to it.
+* Each queue must meet `Allocated >= Used`, it means the queue won't be overused.
+* `Preempting` of each queue must be empty. If it is not empty, it means some pods are terminated to release resources to this queue, but it is not finished.
+
+####Brief workflow:
+
+* Preprocess of `QueueInfo`. Currently, it terminates pods of overused(`Allocated < Used`) queue without preemption. Deserved/Allocated/Used/Preempting will not be adjusted in this stage.
+* Adjust Deserved/Allocated/Used/Preempting of each Queue to trigger preemption, then update results to each Queue.
+* Terminate running pods for each Queue which need preemption, and update Allocated of Queue after preempted pod is finished.
+
+```go
+type Interface interface {
+	// Run start pod informer to handle terminating pods
+	Run(stopCh <-chan struct{})
+
+	// Preprocessing preprocess for each queue
+	// Currently, it terminates pods of overused(Allocated < Used) queue without preemption
+	// Deserved/Allocated/Used/Preempting will not be adjusted
+	Preprocessing(queues map[string]*schedulercache.QueueInfo, pods []*schedulercache.PodInfo) (map[string]*schedulercache.QueueInfo, error)
+
+	// PreemptResources preempt resources between queue
+	// Deserved/Allocated/Used/Preempting will be adjusted
+	PreemptResources(queues map[string]*schedulercache.QueueInfo) error
+}
+```
+
+### Queuejob
+The proportion policy just used Queue `weight` to allocate resources now,  it does not consider the queue job information, such as priority, resources requirement or some other factors. And we need provide some complex scheduling strategy to consider these queue job information.
```

Comment:
Would the scheduling strategy for queuejob be addressed in an additional separate design doc?

Reply:
yes, this is just for API design, we will provide a separate design doc for more scheduling strategy, and I think some discussion may be needed before it.

Reply:
got it.